validator/validate_gamefile.js

```diff
@@ -0,0 +1,879 @@
+const validators = {
+    asset: v=>!!String(v).match(/^\/assets\/-?[0-9]+_[0-9]+$|^\/i\/|^http/) || 'asset expected (format: /assets/1_1, /i/icon.png or http://example.com/image.png)',
+    routineProperty: v=>!!String(v).match(/.Routine$/) || 'routine name expected (format: myRoutine)',
+    idArray: (v,p)=>asArray(v).every(id=>p.widgets[id]) || `widgets ${asArray(v).filter(id=>!p.widgets[id]).join(', ')} not found`,
+    id: (v,p)=>p.widgets[v] || `widget '${v}' not found`,
+    number: v=>typeof v === 'number' || 'number expected',
+    object: v=>typeof v === 'object' && v !== null || 'object expected',
+    boolean: v=>typeof v === 'boolean' || 'boolean expected',
+    string: v=>typeof v === 'string' || 'string expected',
+    inCollection: (v,p)=>{
+        if(p.validCollections[v])
+            return true;
+        if(typeof v === 'string' && p.widgets[v])
+            return `Collection '${v}' not found but it is a widget ID. Did you mean ['${v}'] as an anonymous collection?`;
+        if(typeof v === 'string')
+            return `Collection '${v}' is undefined.`;
+        if(Array.isArray(v) && v.every(id=>p.widgets[id]))
+            return true;
+        if(Array.isArray(v) && v.length === 1 && p.validCollections[v[0]])
+            return `Widget '${v[0]}' does not exist but it is a valid collection. Did you mean '${v[0]}' (without brackets)?`;
+        return false;
+    },
+    routine: (v,p,propertyPath=[])=>{
+        const context = { widgetId: p.widgetId, widgets: p.widgets, validVariables: p.validVariables || { activeColors: 1, mouseCoords: 1, seatIndex: 1, seatID: 1, activeSeats: 1, playerName: 1, playerColor: 1, activePlayers: 1, thisID: 1}, validCollections: p.validCollections || { playerSeats: 1, activeSeats: 1, thisButton: 1 } };
+        return validateRoutine(v,context,propertyPath);
+    },
+    positiveNumber: v=>typeof v === 'number' && v >= 0 || 'positive number expected',
+    property: (v,p)=>Object.values(WIDGET_PROPERTIES).some(props=>Object.keys(props).includes(v)) || Object.values(p.widgets).some(w=>w[v] !== undefined) || `property '${v}' not found`,
+    vttSymbol: v=>v === null || typeof v === 'string', // TODO: replace with actual VTT symbol name check if available
+    countOrAll: v=>v === 'all' || typeof v === 'number' || 'number or "all" expected',
+    any: v=>true
+}
+
+// Common properties for all widgets
```

Comment:
Same comment here as I made below.  Difficult to keep this in sync with the things in the individual widget files when we make changes.